app/controllers/checkout_controller.rb

```diff
@@ -0,0 +1,237 @@
+
+    when :complete
+      order_status = OrderStatus.find_by(status: 'Waiting for Processing')
+      @order.update(order_items: @cart.order_items)
+      @order.update(order_status: order_status)
+    end
+  end
+
+  def update_order_step
+    @order.update(aasm_state: @checkout.current_step)
+  end
+
+  def initialize_order
+    user = current_user
+    order = user.orders.in_progress.first
+
+    if order
+      @order = order
+    else
+      order_status = OrderStatus.find_by(status: 'In Progress')
+      @order = Order.create(user: user, order_status: order_status,
+                            aasm_state: 'address')
+    end
+  end
+
+  def set_shipping_and_billing_address
+    set_shipping_address
+    set_billing_address
+  end
+
+  def set_shipping_address
+    user_info = current_user.user_info
+    begin
+      @shipping_a = ShippingAddress.find(user_info.shipping_address_id)
+    rescue NoMethodError
+      @shipping_a = @order.shipping_address || ShippingAddress.new
+    end
+  end
+
+  def set_billing_address
+    user_info = current_user.user_info
+    begin
+      @billing_a = BillingAddress.find(user_info.billing_address_id)
+    rescue NoMethodError
+      @billing_a = @order.billing_address || BillingAddress.new
+    end
+  end
+
+  def initialize_checkout
+    @checkout = Checkout.new(checkout_params)
+    begin
+      delivery = Delivery.find(@checkout.delivery_method)
+    rescue ActiveRecord::RecordNotFound
+      delivery = @order.delivery
+    end
+
+    @checkout.delivery_method = delivery
+    @checkout.current_step = @order.aasm_state.to_sym
+  end
+
+  def initialize_cart
+    @checkout.cart = @cart
+  end
+
+  def update_cart
+    order_items_ids = "(#{@cart.order_items.ids.join(',')})"
+    @cart.order_items.where("id IN #{order_items_ids}").update_all(cart_id: nil)
+    @cart.update(discount: nil)
+  end
+
+  def checkout_params
+    params.fetch(:checkout, {})
+          .permit(:shipping_first_name, :shipping_last_name,
+                  :shipping_address, :shipping_city, :shipping_zip,
+                  :shipping_country, :shipping_phone,
+                  :billing_first_name, :billing_last_name,
+                  :billing_address, :billing_city, :billing_zip,
+                  :billing_country, :billing_phone, :delivery_method,
+                  :card_number, :cvv, :mm_yy, :use_billing_address, :current_step)
+  end
+
+  def user_params
+    params.require(:user).permit(:email)
+  end
+
+  def set_order_cart
+    @order_cart ||= @cart
+  end
+
+  def redirect_to_login_unless_user_logged_in
+    redirect_to checkout_login_path unless user_signed_in?
+  end
+
+  def redirect_to_catalog_if_cart_empty
+    redirect_to catalog_path if @cart.empty?
+  end
+
+  # == Devise custom form ==
+
+  def resource_name
+    :user
+  end
+
+  def resource
+    @resource ||= User.new
+  end
+
+  def resource_class
+    User
+  end
+
+  def devise_mapping
+    @devise_mapping ||= Devise.mappings[:user]
+  end
+end
```

Comment:
Looks ugly, try to decompose logic, pay attention to `update_order`

Reply:
Need some help with refactoring, especially with `update_order` method.